### PR TITLE
feat: emit `new Error(ID, {cause: CAML_EXN})` for exceptions

### LINKED ISSUE #975: Pass Error messages through to throw new Error
When I raise an exception from my code with a string, I would like to see that string get passed through to the compiled error that is thrown/handled in the compiled js.

Example:

I have this function in melange/ocaml:

```ocaml
let copy_file ~dest file_path =
  try Fs_extra.copySync file_path dest
  with exn ->
    raise
      (Fs_extra_error
         (Printf.sprintf {js|Failed to copy file %s to %s: %s|js} file_path dest
            (Printexc.to_string exn)))
;;
```

and it compiles to the following:

```javascript
function copy_file(dest, file_path) {
  try {
    FsExtra.copySync(file_path, dest);
    return ;
  }
  catch (raw_exn){
    var exn = Caml_js_exceptions.internalToOCamlException(raw_exn);
    throw {
          RE_EXN_ID: Fs_extra_error,
          _1: Curry._3(Stdlib__Printf.sprintf(/* Format */{
                    _0: {
                      TAG: /* String_literal */11,
                      _0: "Failed to copy file ",
                      _1: {
                        TAG: /* String */2,
                        _0: /* No_padding */0,
                        _1: {
                          TAG: /* String_literal */11,
                          _0: " to ",
                          _1: {
                            TAG: /* String */2,
                            _0: /* No_padding */0,
                            _1: {
                              TAG: /* String_literal */11,
                              _0: ": ",
                              _1: {
                                TAG: /* String */2,
                                _0: /* No_padding */0,
                                _1: /* End_of_format */0
                              }
                            }
                          }
                        }
                      }
                    },
                    _1: "Failed to copy file %s to %s: %s"
                  }), file_path, dest, Stdlib__Printexc.to_string(exn)),
          Error: new Error()
        };
  }
}
```

Which results in an error (if not caught and handled) that renders like this: 

image

I would like see my error message passed through to `Error: new Error(mesage)`

### PR DESCRIPTION
**Pull request by anmonteiro**
### Summary

Converts Melange-generated error throwing code to always emit a JS `Error` instance.

fixes #975

#### Before:

```js
      throw {
            MEL_EXN_ID: "Assert_failure",
            _1: [
              "jscomp/test/fun_pattern_match.ml",
              41,
              9
            ],
            Error: new Error()
          };
```

#### After

```js
      throw new Error("Assert_failure", {
                cause: {
                  MEL_EXN_ID: "Assert_failure",
                  _1: [
                    "jscomp/test/fun_pattern_match.ml",
                    41,
                    9
                  ]
                }
              });
```

- Melange now throws an instance of JavaScript's `new Error()` and adds the extension payload for `cause` in the [Error cause option](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause).
- Exception catching runtime was adapted to read `.cause` in the error being caught by `try .. catch(error) { .. }` code

### CODE REVIEW

File: jscomp/runtime/caml_js_exceptions.ml [hunk 1]
```diff
@@ -1,6 +1,7 @@
 open Melange_mini_stdlib
 
 type t = Any : 'a -> t [@@unboxed]
+type 'a js_error = { cause : 'a }
```
[the remark this answers is not in the document]
Comment (jchavarri) [reply jscomp/runtime/caml_js_exceptions.ml:4]: Maybe this type could be `type melange_error = { cause : Caml_exceptions.t }`, I don't understand why the variable has to be added?

**Review by jchavarri — COMMENTED** [review]
Looks great! Some questions for the parts i don't understand.

**Comment by EduardoRFS** [comment]
For both interop and performance reasons, wouldn't it be better to have a MelangeError class? Such that melange errors can be detected with instanceof.

Also is cause supported in older versions or it will be discarded? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Error/cause

Anyway, the MelangeError class could support both `new MelangeError(message, cause)` and it could also be used to ensure `.cause` is defined.

File: jscomp/runtime/caml_js_exceptions.ml (hunk 1 above)
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/runtime/caml_js_exceptions.ml:4]: The variable isn't needed indeed. will change to `type js_error = { cause: exn }`.

This is just a representation of how errors thrown by melange will appear in the generated `catch (error) { ... }` code.

File: jscomp/runtime/caml_js_exceptions.ml [hunk 2]
```diff
@@ -9,7 +10,8 @@ exception Error of t
    [Error] is defined here
 *)
 let internalToOCamlException (e : Obj.t) =
-  if Caml_exceptions.caml_is_extension e then (Obj.magic e : exn)
+  if Caml_exceptions.caml_is_extension (Obj.magic e : _ js_error).cause then
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/runtime/caml_js_exceptions.ml]: Did this in https://github.com/melange-re/melange/pull/1036/commits/b6ff888ddeb422c074a4de32d0488c8c5cc66d97

File: jscomp/core/lam_convert.cppo.ml [hunk 3]
```diff
@@ -24,14 +24,27 @@
...
+    let caml_cause_field_info : Lambda.field_dbg_info =
+      Fld_record { name = Js_dump_lit.cause; mutable_flag = Immutable }
+    in
+    Pfield (0, caml_cause_field_info)
+  in
+  fun loc (head : Lam.t) ->
+    prim
+      ~primitive:lam_caml_id
+      ~args:[
+          (* (prim ~primitive:_lam_caml_cause ~args:[ head ] loc) *)
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/core/lam_convert.cppo.ml]: done in https://github.com/melange-re/melange/pull/1036/commits/01f49b6b013bf7d9ab0dffd8abe17c9532e0604e

File: jscomp/core/lam_convert.cppo.ml [hunk 4]
```diff
@@ -24,14 +24,27 @@
...
 
-let lam_extension_id loc (head : Lam.t) =
-  prim ~primitive:lam_caml_id ~args:[ head ] loc
+let lam_extension_id =
+  let lam_caml_id : Lam_primitive.t =
+    let caml_id_field_info : Lambda.field_dbg_info =
+      Fld_record { name = Js_dump_lit.exception_id; mutable_flag = Immutable }
+    in
+    Pfield (0, caml_id_field_info)
+  and _lam_caml_cause : Lam_primitive.t =
```
[the remark this answers is not in the document]
Comment (anmonteiro) [reply jscomp/core/lam_convert.cppo.ml]: removed in https://github.com/melange-re/melange/pull/1036/commits/01f49b6b013bf7d9ab0dffd8abe17c9532e0604e

**Comment by anmonteiro** [comment]
It might be interesting to play with a `MelangeError` instance too. I'll hold off on that for this PR, but I've opened an issue to track it: https://github.com/melange-re/melange/issues/1042.

**Comment by jchavarri** [comment]
> For both interop and performance reasons, wouldn't it be better to have a MelangeError class?

@EduardoRFS Sorry, I don't understand what would be the upsides of this. In terms of performance, bundling size at least would get worse with `MelangeError` class.

> Such that melange errors can be detected with `instanceof`.

In terms of detection, isn't the `cause` field enough to discern if a given error comes from Melange or not? Or is the proposal to remove `cause` and have it replaced with the new class? (That's not what @anmonteiro is doing in https://github.com/melange-re/melange/pull/1043 is doing, afaict).

**Comment by EduardoRFS** [comment]
@jchavarri I was not talking about bundle size, but even that should shrink, as the MelangeError instance is smaller

```javascript
throw new Error("Assert_failure", {
  cause: {
    MEL_EXN_ID: "Assert_failure",
    _1: [
      "jscomp/test/fun_pattern_match.ml",
      41,
      9
    ]
  }
});

throw new MelangeError("Assert_failure", [
  "jscomp/test/fun_pattern_match.ml",
  41,
  9
]);
```

**Comment by jchavarri** [comment]
Thanks for clarifying. The part that I was missing is that the object with `cause` property would go away.

**Comment by EduardoRFS** [comment]
@jchavarri the object may still be there, it can still be accessed on the instance, is just that it is defined only inside of the constructor.

**Comment by anmonteiro** [comment]
> @EduardoRFS Sorry, I don't understand what would be the upsides of this. In terms of performance, bundling size at least would get worse with `MelangeError` class.

some of the upsides:

- users outside melange can detect whether Melange threw an error with `e instanceof Caml_js_exceptions.MelangeError`
- It also gives us a bit more browser coverage. Check [caniuse](https://caniuse.com/mdn-javascript_builtins_error_cause) for `cause`. 


Improvements in bundling size might be a bit negligible IMO (check https://github.com/melange-re/melange/pull/1043#issuecomment-1917737307) and not the main motivator for this.

**Pull request merged**